### PR TITLE
Initial Hidpi Support

### PR DESCRIPTION
**Pull request by anuejn**
This adds support for scaling on HiDPI screens.
I am unsure if this is the optimal path, but it certainly seemed to be an easy one

### CODE REVIEW

**Comment by anuejn** [comment]
closes #13

File: examples/blur.rs [hunk 1]
```diff
@@ -32,20 +32,22 @@ pub fn draggable_rect(initial_pos: Vec2, context: &mut WidgetContext) -> Fragmen
 #[widget]
 pub fn top(context: &mut WidgetContext) -> Fragment {
     rsx! {
-        <stack>
-            <backdrop_blur sigma=10.>
                 <stack>
-                    <positioned><rect fill=Some(Color::new(1.0, 0.0, 0.0, 1.0))/></positioned>
-                    <text size=200.>{"test text"}</text>
+        //            <backdrop_blur sigma=10.>
```
Comment (anuejn) [inline examples/blur.rs:36]: does this change really belong to this commit?

File: narui_core/src/vulkano_render/lyon.rs [hunk 2]
```diff
@@ -77,11 +77,12 @@ impl<'a> GeometryBuilder for NaruiGeometryBuilder<'a> {
...
     }
 
-    fn end_geometry(&mut self) -> Count {
-        Count {
-            vertices: self.data.vertices.len() as u32 - self.vertex_offset,
-            indices: self.data.indices.len() as u32 - self.index_offset,
-        }
+    fn end_geometry(&mut self) // -> Count
+    {
+        // Count {
```
Comment (anuejn) [inline narui_core/src/vulkano_render/lyon.rs:82]: lets not uncomment this?

File: narui_core/src/vulkano_render/vk_util.rs [hunk 3]
```diff
@@ -38,13 +30,16 @@ impl VulkanContext {
...
-            &instance,
-            MessageSeverity { error: true, warning: true, information: true, verbose: true },
-            MessageType::all(),
-            |msg| log::info!("{}: {}", msg.layer_prefix.unwrap_or("unknown"), msg.description),
-        ));
+        let instance = Instance::new(vulkano::instance::InstanceCreateInfo {
+            enabled_extensions: extensions,
+            ..Default::default()
+        })?;
+        // std::mem::forget(DebugCallback::new(
```
Comment (anuejn) [inline narui_core/src/vulkano_render/vk_util.rs:37]: Isnt this a handy feature?

I would go for 
```
unsafe {
            std::mem::forget(DebugUtilsMessenger::new(
                instance,
                DebugUtilsMessengerCreateInfo {
                    message_severity: DebugUtilsMessageSeverity::all(),
                    message_type: DebugUtilsMessageType::all(),
                    ..DebugUtilsMessengerCreateInfo::user_callback(Arc::new(|msg| log::info!("{}: {}", msg.layer_prefix.unwrap_or("unknown"), msg.description)))
                }
            ));
        }
```

File: narui_core/src/vulkano_render/vk_util.rs [hunk 4]
```diff
@@ -57,16 +52,25 @@ impl VulkanContext {
...
-                let queue_family = physical.queue_families().map(|qf| (qf, 0.5)); // All queues have the same priority
+                let queue_family = physical.queue_families().map(QueueCreateInfo::family).collect(); // All queues have the same priority
                 let device_ext = DeviceExtensions {
                     khr_swapchain: true,
                     khr_storage_buffer_storage_class: true,
                     khr_8bit_storage: true,
                     // Comment in if you need shader printf
                     // khr_shader_non_semantic_info: true,
-                    ..(*physical.required_extensions())
+                    ..DeviceExtensions::none() //                    ..(*physical.required_extensions())
```
Comment (anuejn) [inline narui_core/src/vulkano_render/vk_util.rs:62]: cleanup

File: narui_widgets/src/subpass.rs [hunk 5]
```diff
@@ -224,15 +223,17 @@ pub fn raw_blur(
...
-                let mut set_builder = PersistentDescriptorSet::start(
-                    pipeline.layout().descriptor_set_layouts()[0].clone(),
-                );
-                set_builder
-                    .add_sampled_image(depth, sampler.clone())
-                    .unwrap()
-                    .add_sampled_image(color, sampler.clone())
-                    .unwrap();
-                let descriptor_set = set_builder.build().unwrap();
+                // dbg!(pipeline.layout().set_layouts());
```
Comment (anuejn) [inline narui_widgets/src/subpass.rs:226]: cleanup

File: narui_widgets/src/subpass.rs [hunk 6]
```diff
@@ -224,15 +223,17 @@ pub fn raw_blur(
...
+                let descriptor_set = PersistentDescriptorSet::new(
+                    pipeline.layout().set_layouts()[0].clone(),
+                    [
+                        // TODO(robin): as long as it is unused it gets optimized out??
```
Comment (anuejn) [inline narui_widgets/src/subpass.rs:230]: what does this mean?

**Review by anuejn — COMMENTED** [review]
should I do the cleanup tasks?

**Review by rroohhh — COMMENTED** [review]
Looks fine to me from a quick look, but I did not go through every codepath and check if the sizes / offsets are all in the right units now.

Maybe now (or at some point further down the line) it would be nice to encode which "unit" the sizes / offsets are in into the type system, so you cannot miss conversions?

**Comment by anuejn** [comment]
sounds like a good idea

File: narui_core/src/vulkano_render/lyon.rs (hunk 2 above)
Comment (rroohhh) [reply narui_core/src/vulkano_render/lyon.rs:82]: well there is no return value, I am not really sure what one is supposed to do in this function....

File: narui_core/src/vulkano_render/vk_util.rs (hunk 3 above)
Comment (rroohhh) [reply narui_core/src/vulkano_render/vk_util.rs:37]: ah, that was the one thing I forgot to port

File: narui_core/src/vulkano_render/vk_util.rs (hunk 4 above)
Comment (rroohhh) [reply narui_core/src/vulkano_render/vk_util.rs:62]: right I was not sure if this is the proper way now, do we not need some extensions from the physical device?

File: narui_core/src/vulkano_render/lyon.rs (hunk 2 above)
Comment (anuejn) [reply narui_core/src/vulkano_render/lyon.rs:82]: Probably just delete it?

File: narui_widgets/src/subpass.rs (hunk 6 above)
Comment (rroohhh) [reply narui_widgets/src/subpass.rs:230]: well we do not use the depth sampler in shader, but this causes `set_layouts()[0]` to not contain a entry for it, so if we add the depth sampler here it errors out. Arguably this is a vulkano bug? I did not really investigate further...

File: narui_core/src/vulkano_render/vk_util.rs (hunk 4 above)
Comment (anuejn) [reply narui_core/src/vulkano_render/vk_util.rs:62]: required extensions are now requested automatically so it is safe to just delete that line

**Comment by rroohhh** [comment]
Ok, then I think it would be nice to do the encoding into the typesystem already in this PR.

Not sure yet, but it would also probably be nice for consumers (people using measurement hooks or so) to get pixel and scaled pixel size maybe? (could be useful to decide on which resolution of a icon to use or something like that)

**Comment by rroohhh** [comment]
@anuejn please review

File: narui_core/src/eval/layout.rs [hunk 7]
```diff
@@ -8,8 +8,65 @@ use derivative::Derivative;
 use freelist::Idx;
 use rutter_layout::{layouter::LayoutIterDirection, BoxConstraints, Layout, Offset};
 use std::ops::Deref;
+use winit::dpi::PhysicalPosition;
 
-/// PositionedRenderObject is the main output data structure of the Layouting
+// type to create a seperation between logical and physical pixels
```
Comment (anuejn) [inline narui_core/src/eval/layout.rs:13]: I like this design :)

**Review by anuejn — COMMENTED** [review]
I cannot approve this (since I opened this PR) but I am fine with the changes :)

**Review by rroohhh — APPROVED** [review]

**Pull request merged**